internal/app/machined/internal/sequencer/v1alpha1/v1alpha1_sequencer.go

```diff
@@ -28,221 +30,266 @@ import (
-	}
+func (s *Sequencer) Shutdown() error {
+	f := func() error {
+		content, err := config.FromFile(constants.ConfigPath)
+		if err != nil {
+			return err
+		}
 
-	config, err := config.New(content)
-	if err != nil {
-		return err
-	}
+		config, err := config.New(content)
+		if err != nil {
+			return err
+		}
 
-	phaserunner, err := phase.NewRunner(config, runtime.Shutdown)
-	if err != nil {
-		return err
-	}
+		phaserunner, err := phase.NewRunner(config, runtime.Shutdown)
+		if err != nil {
+			return err
+		}
 
-	phaserunner.Add(
-		phase.NewPhase(
-			"stop services",
-			services.NewStopServicesTask(false),
-		),
-	)
+		phaserunner.Add(
+			phase.NewPhase(
+				"stop services",
+				services.NewStopServicesTask(false),
+			),
+		)
 
-	return phaserunner.Run()
+		return phaserunner.Run()
+	}
+
+	return s.run(f)
 }
 
 // Upgrade implements the Sequencer interface.
-func (d *Sequencer) Upgrade(req *machineapi.UpgradeRequest) error {
-	content, err := config.FromFile(constants.ConfigPath)
-	if err != nil {
-		return err
-	}
+func (s *Sequencer) Upgrade(req *machineapi.UpgradeRequest) error {
+	f := func() error {
+		time.Sleep(10 * time.Second)
+		content, err := config.FromFile(constants.ConfigPath)
+		if err != nil {
+			return err
+		}
 
-	config, err := config.New(content)
-	if err != nil {
-		return err
-	}
+		config, err := config.New(content)
+		if err != nil {
+			return err
+		}
 
-	phaserunner, err := phase.NewRunner(config, runtime.Upgrade)
-	if err != nil {
-		return err
+		phaserunner, err := phase.NewRunner(config, runtime.Upgrade)
+		if err != nil {
+			return err
+		}
+
+		var dev *probe.ProbedBlockDevice
+
+		dev, err = probe.GetDevWithFileSystemLabel(constants.EphemeralPartitionLabel)
+		if err != nil {
+			return err
+		}
+
+		devname := dev.Device().Name()
+
+		if err := dev.Close(); err != nil {
+			return err
+		}
+
+		phaserunner.Add(
+			phase.NewPhase(
+				"cordon and drain node",
+				kubernetes.NewCordonAndDrainTask(),
+				upgrade.NewLeaveEtcdTask(),
+			),
+			phase.NewPhase(
+				"remove all pods",
+				kubernetes.NewRemoveAllPodsTask(),
+			),
+			phase.NewPhase(
+				"stop services",
+				services.NewStopServicesTask(true),
+			),
+			phase.NewPhase(
+				"unmount system disk submounts",
+				rootfs.NewUnmountOverlayTask(),
+				rootfs.NewUnmountPodMountsTask(),
+			),
+			phase.NewPhase(
+				"unmount system disk",
+				rootfs.NewUnmountSystemDisksTask(devname),
+			),
+			phase.NewPhase(
+				"verify system disk not in use",
+				disk.NewVerifyDiskAvailabilityTask(devname),
+			),
+			phase.NewPhase(
+				"reset system disk",
+				disk.NewResetSystemDiskTask(devname),
+			),
+			phase.NewPhase(
+				"upgrade",
+				upgrade.NewUpgradeTask(devname, req),
+			),
+		)
+
+		return phaserunner.Run()
 	}
 
-	var dev *probe.ProbedBlockDevice
+	return s.run(f)
+}
 
-	dev, err = probe.GetDevWithFileSystemLabel(constants.EphemeralPartitionLabel)
-	if err != nil {
-		return err
+func (s *Sequencer) Locked() bool {
```

Comment:
:+1: :zap: :cool: 

idk, if method name is self-descriptive, or should it be `TryLock` or something? (nit)

Reply:
yeah that sounds way better ... I'm all for good naming